### ISSUE TITLE
RC processing machine code runs on client, item transport only serverside, causing desync

### ISSUE THREAD

**Issue by Tyler799** [ReikaKalseki/Reika_Mods_Issues#1678]
Rotarycraft version: v19c

With a gasoline engine, the Pneumatic Item pump transfers 1 stack per tick.

In testing between two JABBA barrels, this is exactly the case.

However, when transferring from a JABBA barrel to a Centrifuge, this doesn't seem to work. I'm not sure exactly of the rate, but it transfers in groups of 10 items at a time, in about a little under a second each.

This particular centrifuge is operating at 1 megarads/second, or 20 items per second.

The Pneumatic Item pump should be _more_ than able to keep up with it, but it is not due to this issue.

**Comment by ReikaKalseki** [comment]
It might have a built-in throttle behavior if it cannot fit a full stack in the target inventory. I would need to double-check.

**Comment by Tyler799** [comment]
Hey. Quick update. I did more testing. It's not the fault of the Pneumatic Item Pump. It tested with a Pneumaticraft omnidirectional hopper. It looks like even when it tries to transfer 256 items/tick into the centrifuge, the centrifuge just refuses to accept them more than about 10/second or so. (Which isn't fast enough to keep up with 1-tick operation).

So the issue is with the centrifuge, not the pump.

**Comment by ReikaKalseki** [comment]
Are you sure? I have four 3-operations-per-tick centrifuges maintained at 100% full.

**Comment by Tyler799** [comment]
How can you have a "3 operation per tick" centrifuge? The minimum operation time of any machine in rotarycraft is 1 tick (0.05s), right? 

In any case, I'm positive. I tried both the pneumatic item pump with a gasoline engine (1 stack per tick) and the omnidirectional hopper with 256 items/tick.

For some reason, the centrifuge just wasn't letting in items fast enough. It was letting in groups of about 10 every second or so. So it always had an empty inventory. 

Could it be due to the side I was inputting from? I input from a side and not the top, if that matters.

EDIT: Tried top, that didn't work either. Also tried several pneumatic item pumps all powered by gasoline engines on different sides. Centrifuge still refuses to take in items fast enough.

**Comment by Tyler799** [comment]
I did some more testing and made a video with the specifics.

It may just be a display issue after all.

Here: https://youtu.be/g1MaexJEGI4

**Comment by ReikaKalseki** [comment]
> How can you have a "3 operation per tick" centrifuge? The minimum operation time of any machine in rotarycraft is 1 tick (0.05s), right?

No. As of v8, many machines support multiple operations per tick if given even *more* speed.

> It may just be a display issue after all.

Is it a client/server issue? Try rapidly reopening the GUI; does that resolve the issue?

**Comment by Tyler799** [comment]
If you watch the video, that's exactly what I did at the end.

Every time I open the gui, it appears to be full, then drain completely and be empty. 

If I rapidly open it, it always "appears" full. 

Also, I've been frequently updating the main 3 wikis for Rotarycraft as I come across newer information.

Which machines support multiple operations per tick? Every single wiki, every single resources I've found says 1 tick operation is as fast as it goes. I'd like to set that straight.

1200 - 60 × log2 (speed + 1) is the formula for the centrifuge's duration time.

At 1 megarads, it operates in -0.00083 ticks, which seems to come out to 1 tick rounded up.
At 0.5 megarads, it operates in about 60 ticks, or 3 seconds or so.

So what constitutes 2 operations per tick? Or more?

**Comment by ReikaKalseki** [comment]
So then yes, the issue is that the GUI contents are not being resynced properly, and since the item transport code only runs on the server side, while the processing runs on both sides, you get this.

> Which machines support multiple operations per tick?

Anything implementing https://github.com/ReikaKalseki/RotaryCraft/blob/master/Auxiliary/Interfaces/MultiOperational.java

> So what constitutes 2 operations per tick? Or more?

https://github.com/ReikaKalseki/RotaryCraft/blob/master/Auxiliary/DurationFormula.java

https://github.com/ReikaKalseki/RotaryCraft/blob/master/Registry/DurationRegistry.java

**Comment by Tyler799** [comment]
So the issue is just the gui, not the machine. Good to know. Thanks. I suppose you can close this now.

BTW, I meant to ask. Is it intended that both the Microturbine and Gas turbine are twice as inefficient at 0.0625 (idle) speed with an ECU? I just did some testing. They consume 12.5% of the fuel they would at full speed, despite running at only 6.25% of the speed.

I'm assuming it's intentional since the numbers round out so nicely to 2x.

**Comment by ReikaKalseki** [comment]
Yes. Turbine engines do not throttle nearly as well.
